internal/update.go

```diff
@@ -0,0 +1,354 @@
+func (up *updateProtocol) Reject(err error) {
+	up.requireState("reject", updateStateNew, updateStateRequestInitiated)
+	up.env.Send(&protocolpb.Message{
+		ProtocolInstanceId: up.protoInstanceID,
+		Body: protocol.MustMarshalAny(&updatepb.Rejection{
+			RejectedRequestMessageId:         up.requestMsgID,
+			RejectedRequestSequencingEventId: up.requestSeqID,
+			RejectedRequest:                  &up.initialRequest,
+			Failure:                          up.env.GetFailureConverter().ErrorToFailure(err),
+		}),
+	})
+	up.state = updateStateCompleted
+}
+
+// Complete is called for an update with the result of executing the
+// update function.
+func (up *updateProtocol) Complete(success interface{}, outcomeErr error) {
+	up.requireState("complete", updateStateAccepted)
+	outcome := &updatepb.Outcome{}
+	if outcomeErr != nil {
+		outcome.Value = &updatepb.Outcome_Failure{
+			Failure: up.env.GetFailureConverter().ErrorToFailure(outcomeErr),
+		}
+	} else {
+		success, err := up.env.GetDataConverter().ToPayloads(success)
+		if err != nil {
+			panic(err)
+		}
+		outcome.Value = &updatepb.Outcome_Success{
+			Success: success,
+		}
+	}
+	up.env.Send(&protocolpb.Message{
+		ProtocolInstanceId: up.protoInstanceID,
+		Body: protocol.MustMarshalAny(&updatepb.Response{
+			Meta:    up.initialRequest.GetMeta(),
+			Outcome: outcome,
+		}),
+	})
+	up.state = updateStateCompleted
+}
+
+// defaultHandler receives the initial invocation of an upate during WFT
+// processing. The implementation will verify that an updateHandler exists for
+// the supplied name (rejecting the update otherwise) and use the provided spawn
+// function to create a new coroutine that will execute in the workflow context.
+// The spawned coroutine is what will actually invoke the user-supplied callback
+// functions for validation and execution. Update progress is emitted via calls
+// into the UpdateCallbacks parameter.
+func defaultUpdateHandler(
+	rootCtx Context,
+	name string,
+	serializedArgs *commonpb.Payloads,
+	header *commonpb.Header,
+	callbacks UpdateCallbacks,
+	spawn func(Context, string, func(Context)) Context,
+) {
+	env := getWorkflowEnvironment(rootCtx)
+	ctx, err := workflowContextWithHeaderPropagated(rootCtx, header, env.GetContextPropagators())
+	if err != nil {
+		callbacks.Reject(err)
+		return
+	}
+	eo := getWorkflowEnvOptions(ctx)
+	handler, ok := eo.updateHandlers[name]
+	if !ok {
+		keys := make([]string, 0, len(eo.updateHandlers))
+		for k := range eo.updateHandlers {
+			keys = append(keys, k)
+		}
+		callbacks.Reject(fmt.Errorf("unknown update %v. KnownUpdates=%v", name, keys))
+		return
+	}
+
+	args, err := decodeArgsToRawValues(
+		env.GetDataConverter(),
+		reflect.TypeOf(handler.fn),
+		serializedArgs,
+	)
+	if err != nil {
+		callbacks.Reject(fmt.Errorf("unable to decode the input for update %q: %w", name, err))
+		return
+	}
+	input := UpdateInput{Name: name, Args: args}
+
+	spawn(ctx, name, func(ctx Context) {
```

Comment:
Reminder: We'll want SDK features tests for panic in validation handler and panic in update. Obviously with Go code they are the same (wft failure), but it forces other SDKs to acknowledge their unexpected-error paths too.

Reply:
Sure. And FYI they're in the unit tests too.